### ISSUE TITLE
Check if call() is type compatible with arguments

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#130]
Reference: https://youtrack.jetbrains.com/issue/WI-34954

For some reason, PS have some issues with the "type compatibility" inspection, and I guess that you can solve that with your extension.

Basically, if you declare a method like `test($arg)` and PHPDoc with `bool $arg`, it'll be reported if I call test like `test(new stdClass)`, but will do not if I do with incompatible type, like `test(123)`. It's worse if I declare function with a default value like `test($arg = null)`, then it now will `test(new stdClass)`.

```php
/** @param bool $arg */
function test($arg) {}

test(new stdClass); // Expected: REPORT, Result: REPORT (OK)
test(123); // Expected: REPORT, Result: NOT REPORT (FAIL)

/** @param bool $arg */
function test($arg = null) {}

test(new stdClass); // Expected: REPORT, Result: NOT REPORT (FAIL)
test(123); // Expected: REPORT, Result: NOT REPORT (FAIL)
```

Then my suggestion is: create a visitor to call()'s to check if type passed is compatible with called function/method. 

---

**Important case:** there a case where you have a bit more atention about "subtypes". I guess that it should be applied only for boolean, where you can PHPDoc it only as "false" or "true" (and not "bool" to both). Then if I have a PHPDoc like `true|null $arg`, it should report if I do `call(false)`.

**Comment by funivan** [comment]
@rentalhost have same idea. Would be nice to have this check. Also this can cover some NPE:
```php
function a() : ?string {
}
function b(string $value){
}
b(a()) << error: null check
```

@kalessil I can try to implement this, don't know how, but want to try))

**Comment by kalessil** [comment]
@funivan : I'd skip what related to PhpDoc. In case of implicit type hints, it'd have more sense, but I'm not sure if this in demand without strict types enabled (then IDE starts reporting the mismatches).

What would be cool is to find a way to make an inspection which enables the mentioned functionality without strict types directive: that would be super-welcome.

So if you want to give a try to implement this super-welcome feature, be my guest =)
